### ISSUE TITLE
Big performance regression with new `FermionicOp`

### ISSUE THREAD

**Issue by jlapeyre** [Qiskit/qiskit-nature#234]
I am using this code:
```python

from qiskit_nature.drivers import PySCFDriver, UnitsType
from qiskit_nature.problems.second_quantization.electronic.builders.fermionic_op_builder import build_ferm_op_from_ints
from qiskit_nature.operators.second_quantization import FermionicOp

h2o = "O 0.0 0.0 0.0;H 0.757 0.586 0.0;H -0.757 0.586 0.0"
h2 = 'H .0 .0 0.7414; H .0 .0 .0'
lih = 'Li .0 .0 0.7414; H .0 .0 .0'
basis0 = "sto3g"
basis1 = "631g"

def get_molecule():

    driver = PySCFDriver(atom=h2o,
                         unit=UnitsType.ANGSTROM,
                         charge=0,
                         spin=0,
                         basis=basis0)
    molecule = driver.run()
    return molecule

def get_qkn_op(molecule):
    one_body = molecule.one_body_integrals
    two_body = molecule.two_body_integrals
    fop_qkn = build_ferm_op_from_ints(one_body, two_body)
    return fop_qkn

molecule = get_molecule()
```
Then, after the change of internals, with a0f4c7bb , I get this:
```python
%time fop = get_qkn_op(molecule)
CPU times: user 12.2 s, sys: 58.4 ms, total: 12.3 s
Wall time: 4.87 s
```

However, before the rewrite of internals 9f75adc, I get this:
```python
%time fop = get_qkn_op(molecule)
CPU times: user 316 ms, sys: 1.83 ms, total: 317 ms
Wall time: 318 ms
```

If I use `basis1` (631g) above, then the time before the change is 4s. The time after the change is longer than several minutes; I killed the process.

**Comment by mrossinek** [comment]
I can confirm that 8645ee35fa40d1c4d67d71e8a1c0e766f8af324b is the first commit on `main` which experiences this drastic performance degradation. @ikkoham could you please also take a look at this one?

**Comment by ikkoham** [comment]
hmm... Coexistence of dense and sparse labels causes the performance regression. I think we should separate the internal structure and API as I first proposed. But, I will try to improve the performance when I do #224 .

**Comment by jlapeyre** [comment]
@mrossinek . I did not identify the first release that introduces the regression. I assumed it was the introduction of the new structure.

**Comment by mrossinek** [comment]
> Coexistence of dense and sparse labels causes the performance regression

I agree. But I didn't expect this to be >10x slowdown..

**Comment by ikkoham** [comment]
It's slow because it's converting between dense and sparse. It's true that my implementation was rushed and very bad. There are a lot of unnecessary conversions, so I thought I would fix the conversions when I delete Legacy.

**Comment by jlapeyre** [comment]
I didn't look into this at all, just reported it. The reason is that I was guessing that the response might be that my use of `build_ferm_op_from_ints` is not recommended, i.e. is obsolete.  Of course sending a sparse object to routines optimized for dense storage or vice versa often incurs a big performance penalty due to conversion. I suppose functions will need to be rewritten to take advantage of the sparse structure.

**Comment by mrossinek** [comment]
Closed via #314
